File: kamacoder53/kamacoder53.cpp
// prim算法

// #include<iostream>
// #include<vector>
// #include<climits>

// using namespace std;

// int main(){
//     int v, e;
//     int x, y, k;
//     cin >> v >> e;
    
//     vector<vector<int>> grid(v + 1, vector<int>(v + 1, 10001));
//     while (e--)
//     {
//         cin >> x >> y >> k;
//         grid[x][y] = k;
//         grid[y][x] = k;
//     }
    
//     vector<int> minDist(v + 1, 10001);
//     vector<bool> isInTree(v + 1, false);

//     for (int i = 1; i < v; i++){
//         int cur = -1;
//         int minVal = INT_MAX;
//         for (int j = 1; j <= v; j++){
//             if (!isInTree[j] && minDist[j] < minVal){
//                 minVal = minDist[j];
//                 cur = j;
//             }
//         }
//         isInTree[cur] = true;


//         for (int j = 1; j <= v; j++){
//             if (!isInTree[j] && grid[cur][j] < minDist[j]){
//                 minDist[j] = grid[cur][j];
//             }
//         }
//     }
//     int result = 0;
//     for (int i = 2; i <= v; i++){
//         result += minDist[i];
//     }
//     cout << result << endl;
// }


// kruskal算法

#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

struct Edge
{
    int l, r, val;
};
int n = 10001;
vector<int> father(n, -1);

void init(){
    for (int i = 0; i < n; i++){
        father[i] = i;
    }
}

int find(int u){
    return u == father[u] ? u : father[u] = find(father[u]);
}

void join(int u, int v){
    u = find(u);
    v = find(v);
    if (u == v) return;
    father[v] = u;
}

int main(){
    int v, e;
    int v1, v2, val;
    vector<Edge> edges;
    int result_val = 0;
    cin >> v >> e;
    while (e--)
    {
        cin >> v1 >> v2 >> val;
        edges.push_back({v1, v2, val});
    }

    sort(edges.begin(), edges.end(), [](const Edge& a, const Edge& b){
        return a.val < b.val;
    });

    init();

    for (Edge edge : edges){
        int x = find(edge.l);
        int y = find(edge.r);

        if (x != y) {
            result_val += edge.val;
            join(x, y);
        }
    }
    cout << result_val << endl;
    return 0;
}

File: kamacoder53/kamacoder53.py
# prim算法
# v, e = list(map(int, input().strip().split()))

# graph = [[10001] * (v + 1) for _ in range(v + 1)]

# for _ in range(e):
#     x, y, w = list(map(int, input().strip().split()))
#     graph[x][y] = w
#     graph[y][x] = w

# visited = [False] * (v + 1)
# minDist = [10001] * (v + 1)

# for _ in range(1, v + 1):
#     min_val = 10002
#     cur = -1
#     for j in range(1, v + 1):
#         if visited[j] == False and minDist[j] < min_val:
#             cur = j
#             min_val = minDist[j]
#     visited[cur] = True
#     for j in range(1, v + 1):
#         if visited[j] == False and minDist[j] > graph[cur][j]:
#             minDist[j] = graph[cur][j]

# ans = 0
# for i in range(2, v + 1):
#     ans += minDist[i]
# print(ans)

# kruskal

class Edge:
    def __init__(self, l, r, val):
        self.l = l
        self.r = r
        self.val = val

n = 10001
father = list()

def init():
    global father
    father = list(range(n))

def find(u):
    if u != father[u]:
        father[u] = find(father[u])
    return father[u]
def join(u, v):
    u = find(u)
    v = find(v)
    if u != v:
        father[v] = u
def kruskal(v, edges):
    edges.sort(key = lambda edge: edge.val)
    init()
    result_val = 0

    for edge in edges:
        x = find(edge.l)
        y = find(edge.r)
        if x != y:
            result_val += edge.val
            join(x, y)
    return result_val



v, e = list(map(int, input().strip().split()))

edges = []
for _ in range(e):
    v1, v2, val = map(int, input().strip().split())
    edges.append(Edge(v1, v2, val))

result_val = kruskal(v, edges)
print(result_val)